core/clustersmngr/factory.go

```diff
@@ -0,0 +1,230 @@
+package clustersmngr
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/cheshir/ttlcache"
+	"github.com/go-logr/logr"
+	"github.com/weaveworks/weave-gitops/core/nsaccess"
+	"github.com/weaveworks/weave-gitops/pkg/server/auth"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/rest"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+//go:generate go run github.com/maxbrunsfeld/counterfeiter/v6 -generate
+
+const (
+	userNamespaceTTL        = 1 * time.Hour
+	watchClustersFrequency  = 30 * time.Second
+	watchNamespaceFrequency = 30 * time.Second
+)
+
+// ClientsFactory is a factory for creating clients for clusters
+//counterfeiter:generate . ClientsFactory
+type ClientsFactory interface {
+	// GetImpersonatedClient returns the clusters client for the given user
+	GetImpersonatedClient(ctx context.Context, user *auth.UserPrincipal) (Client, error)
+	// UpdateClusters updates the clusters list
+	UpdateClusters(ctx context.Context) error
+	// UpdateNamespaces updates the namespaces all namespaces for all clusters
+	UpdateNamespaces(ctx context.Context) error
+	// GetServerClient returns the cluster client with gitops server permissions
+	GetServerClient(ctx context.Context) (Client, error)
+	// GetClustersNamespaces returns the namespaces for all clusters
+	GetClustersNamespaces() map[string][]v1.Namespace
+	// Start starts go routines to keep clusters and namespaces lists up to date
+	Start(ctx context.Context)
+}
+
+type clientsFactory struct {
+	clustersFetcher ClusterFetcher
+	nsChecker       nsaccess.Checker
+	log             logr.Logger
+
+	// list of clusters returned by the clusters fetcher
+	clusters *Clusters
+	// the lists of all namespaces of each cluster
+	clustersNamespaces *ClustersNamespaces
+	// lists of namespaces accessible by the user on every cluster
+	usersNamespaces *UsersNamespaces
+}
+
+func NewClientFactory(hubClient client.Client, fetcher ClusterFetcher, nsChecker nsaccess.Checker, logger logr.Logger) ClientsFactory {
```

Comment:
```suggestion
func NewClientFactory(fetcher ClusterFetcher, nsChecker nsaccess.Checker, logger logr.Logger) ClientsFactory {
```
this can be removed as an arg now